File: backend/app/api/v1/notifications.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import List

from app.database import get_database
from app.core.dependencies import get_current_user
from app.schemas.notification import (
    NotificationResponse,
    NotificationList,
    NotificationUpdate,
    UnreadCountResponse
)
from app.schemas.common import MessageResponse
from app.services.notification_service import NotificationService
from app.utils.exceptions import NotFoundException


router = APIRouter(prefix="/notifications", tags=["Notifications"])


@router.get("", response_model=NotificationList)
async def get_notifications(
    skip: int = Query(0, ge=0, description="Number of notifications to skip"),
    limit: int = Query(50, ge=1, le=100, description="Number of notifications to return"),
    unread_only: bool = Query(False, description="Return only unread notifications"),
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    Get notifications for the current user.
    
    - **skip**: Pagination offset
    - **limit**: Number of notifications to return (max 100)
    - **unread_only**: Filter to show only unread notifications
    """
    notification_service = NotificationService(db)
    
    notifications, total, unread_count = await notification_service.get_user_notifications(
        user_id=str(current_user["_id"]),
        skip=skip,
        limit=limit,
        unread_only=unread_only
    )
    
    # Convert ObjectId to string
    for notification in notifications:
        notification["_id"] = str(notification["_id"])
    
    return {
        "notifications": notifications,
        "total": total,
        "unreadCount": unread_count
    }


@router.get("/unread-count", response_model=UnreadCountResponse)
async def get_unread_count(
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Get count of unread notifications for the current user."""
    notification_service = NotificationService(db)
    
    count = await notification_service.get_unread_count(
        user_id=str(current_user["_id"])
    )
    
    return {"count": count}


@router.patch("/{notification_id}/read", response_model=NotificationResponse)
async def mark_notification_as_read(
    notification_id: str,
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Mark a specific notification as read."""
    notification_service = NotificationService(db)
    
    try:
        notification = await notification_service.mark_as_read(
            notification_id=notification_id,
            user_id=str(current_user["_id"])
        )
        notification["_id"] = str(notification["_id"])
        return notification
    except NotFoundException as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.patch("/mark-all-read", response_model=MessageResponse)
async def mark_all_as_read(
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Mark all notifications as read for the current user."""
    notification_service = NotificationService(db)
    
    count = await notification_service.mark_all_as_read(
        user_id=str(current_user["_id"])
    )
    
    return {"message": f"Marked {count} notifications as read"}


@router.delete("/{notification_id}", response_model=MessageResponse)
async def delete_notification(
    notification_id: str,
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Delete a specific notification."""
    notification_service = NotificationService(db)
    
    try:
        await notification_service.delete_notification(
            notification_id=notification_id,
            user_id=str(current_user["_id"])
        )
        return {"message": "Notification deleted successfully"}
    except NotFoundException as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.delete("", response_model=MessageResponse)
async def delete_all_notifications(
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Delete all notifications for the current user."""
    notification_service = NotificationService(db)
    
    count = await notification_service.delete_all_notifications(
        user_id=str(current_user["_id"])
    )
    
    return {"message": f"Deleted {count} notifications"}

File: client/src/lib/api/notifications.ts
import axios from 'axios';
import type { NotificationList, Notification, UnreadCountResponse, MessageResponse } from '@/types/api';

const API_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1';

// Helper to get auth header
const getAuthHeader = () => {
  const token = localStorage.getItem('token');
  return { Authorization: `Bearer ${token}` };
};

/**
 * Get notifications for the current user
 */
export async function getNotifications(
  skip: number = 0,
  limit: number = 50,
  unreadOnly: boolean = false
): Promise<NotificationList> {
  const response = await axios.get(`${API_URL}/notifications`, {
    headers: getAuthHeader(),
    params: { skip, limit, unread_only: unreadOnly },
  });
  return response.data;
}

/**
 * Get unread notification count
 */
export async function getUnreadCount(): Promise<UnreadCountResponse> {
  const response = await axios.get(`${API_URL}/notifications/unread-count`, {
    headers: getAuthHeader(),
  });
  return response.data;
}

/**
 * Mark a notification as read
 */
export async function markAsRead(notificationId: string): Promise<Notification> {
  const response = await axios.patch(
    `${API_URL}/notifications/${notificationId}/read`,
    {},
    { headers: getAuthHeader() }
  );
  return response.data;
}

/**
 * Mark all notifications as read
 */
export async function markAllAsRead(): Promise<MessageResponse> {
  const response = await axios.patch(
    `${API_URL}/notifications/mark-all-read`,
    {},
    { headers: getAuthHeader() }
  );
  return response.data;
}

/**
 * Delete a notification
 */
export async function deleteNotification(notificationId: string): Promise<MessageResponse> {
  const response = await axios.delete(`${API_URL}/notifications/${notificationId}`, {
    headers: getAuthHeader(),
  });
  return response.data;
}

/**
 * Delete all notifications
 */
export async function deleteAllNotifications(): Promise<MessageResponse> {
  const response = await axios.delete(`${API_URL}/notifications`, {
    headers: getAuthHeader(),
  });
  return response.data;
}
